File: sdk/go/cmd/vorpal/artifact/vorpal_container_image.go
package artifact

import (
	"fmt"

	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

func BuildVorpalContainerImage(context *config.ConfigContext) (*string, error) {
	linuxVorpalSlim, err := artifact.LinuxVorpalSlim(context)
	if err != nil {
		return nil, fmt.Errorf("failed to build vorpal: %w", err)
	}

	vorpal, err := Vorpal(context)
	if err != nil {
		return nil, fmt.Errorf("failed to build vorpal: %w", err)
	}

	name := "vorpal-container-image"

	return artifact.NewOciImage(name, *linuxVorpalSlim).
		WithAliases([]string{fmt.Sprintf("%s:latest", name)}).
		WithArtifacts([]*string{vorpal}).
		Build(context)
}

File: config/src/artifact/vorpal_container_image.rs
use crate::Vorpal;
use anyhow::Result;
use vorpal_sdk::{
    artifact::{linux_vorpal_slim::LinuxVorpalSlim, oci_image::OciImage},
    context::ConfigContext,
};

#[derive(Default)]
pub struct VorpalContainerImage {}

impl VorpalContainerImage {
    pub fn new() -> Self {
        Self::default()
    }

    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let linux_vorpal_slim = LinuxVorpalSlim::new().build(context).await?;
        let vorpal = Vorpal::new().build(context).await?;

        let name = "vorpal-container-image";

        OciImage::new(name, &linux_vorpal_slim)
            .with_aliases(vec![&format!("{}:latest", name)])
            .with_artifacts(vec![&vorpal])
            .build(context)
            .await
    }
}
